typescript/introduction/tsc-data-types-ii/tsc-object-type-comparison.md

```diff
@@ -0,0 +1,124 @@
+---
+author: kapnobatai136
+
+type: normal
+
+category: must-know
+
+aspects:
+  - introduction
+
+---
+
+# `Object` vs `object` vs `{}`
+
+---
+## Content
+
+TypeScript has 3 ways to represent an object with slightly different behaviors.
+
+Since almost everything in JavaScript is an object[1], the `Object` type can be used to represent any type that is considered to contain a value:
+
+```ts
+let num: Object = 3;
+let str: Object = 'enki';
+let arr: Object = [1, 2, 3];
+// ...
+// only null and undefined give an error
+let n: Object = null; // error
+let u: Object = undefined; // error
+```
+
+This often leads to confusion. Usually by "object" we only mean collections of values, whether it be maps, sets, arrays, etc or plain objects, but not primitive values such as numbers or strings.
+
+This is where the `object` type comes into play.
+
+```ts
+let arr: object = [1, 2, 3];
+let obj: object = { enki: true };
+let map: object = new Map([["enki", "cool"]])
+// ...
+// primitive values give an error
+let str: object = 'enki'; // error
+let num: object = 3; // error
+let n: object = null; // error
+let u: object = undefined; // error
+```
+
+The type `{}` is used to represent an empty JavaScript object.
+
+```ts
+const empty: {} = {};
+```
```

Comment:
````suggestion
The type `{}` is used to represent a JavaScript object with a specific shape.

```ts
let empty: {} = {}

const Enki: {
  job: string,
  isFun: boolean,
} = { job: "teach", isFun: true };
```

Assigning values that aren't part of the shape will cause errors:

```ts
empty.name = 'oops'; // Property 'name' does not exist on type '{}'
Enki.age = 5; // Property 'age' does not exist on type '{ job: string; isFun: boolean; }'
```
````